fletcher/_algorithms.py

```diff
@@ -261,3 +281,187 @@ def all_op(arr: Union[pa.ChunkedArray, pa.Array], skipna: bool) -> bool:
+            chunk for chunk in arr.iterchunks() if chunk.null_count != len(chunk)
+        ]
+        results = pa.array(
+            [
+                np_reduce_op(npop, chunk, skipna=skipna, identity=identity)
+                for chunk in valid_chunks
+            ]
+        )
+        return np_reduce_op(npop, results, skipna=skipna, identity=identity)
+    else:
+        if len(arr) == 0:
+            if identity is None:
+                raise ValueError("zero-size reduction on operation with no identity")
+            else:
+                return identity
+        np_arr = _extract_data_buffer_as_np_array(arr)
+        if arr.null_count > 0:
+            mask = extract_isnull_bytemap(arr)
+            return npop(np_arr[~mask])
+        else:
+            return npop(np_arr)
+
+
+sum_op = partial(np_reduce_op, np.sum, identity=0)
+max_op = partial(np_reduce_op, np.amax)
+min_op = partial(np_reduce_op, np.amin)
+prod_op = partial(np_reduce_op, np.prod, identity=1)
+
+
+def pd_nanop(nanop: Callable, arr: Union[pa.ChunkedArray, pa.Array], skipna: bool):
+    """Use pandas.core.nanops to provide a reduction."""
+    if isinstance(arr, pa.ChunkedArray):
+        data = pa.concat_arrays(arr.iterchunks())
+    else:
+        data = arr
+    np_arr = _extract_data_buffer_as_np_array(data)
+    mask = extract_isnull_bytemap(data)
+
+    return nanop(np_arr, skipna=skipna, mask=mask)
+
+
+std_op = partial(pd_nanop, nanops.nanstd)
+skew_op = partial(pd_nanop, nanops.nanskew)
+kurt_op = partial(pd_nanop, nanops.nankurt)
+var_op = partial(pd_nanop, nanops.nanvar)
+median_op = partial(pd_nanop, nanops.nanmedian)
+
+
+def _in_chunk_offsets(
+    arr: pa.ChunkedArray, offsets: List[int]
+) -> List[Tuple[int, int, int]]:
+    new_offsets = []
+    pos = 0
+    chunk = 0
+    chunk_pos = 0
+    for offset, offset_next in zip(offsets, offsets[1:] + [len(arr)]):
+        diff = offset - pos
+        chunk_remains = len(arr.chunk(chunk)) - chunk_pos
+        step = offset_next - offset
+        if diff == chunk_remains:
+            chunk += 1
+            chunk_pos = 0
+            pos += chunk_remains
+            new_offsets.append((chunk, chunk_pos, step))
+        elif diff == 0:  # The first offset
+            new_offsets.append((chunk, chunk_pos, step))
+        else:  # diff < chunk_remains
+            chunk_pos += diff
+            pos += diff
+            new_offsets.append((chunk, chunk_pos, step))
+    return new_offsets
+
+
+def _combined_in_chunk_offsets(
+    a: pa.ChunkedArray, b: pa.ChunkedArray
+) -> Tuple[List[Tuple[int, int, int]], List[Tuple[int, int, int]]]:
+    offsets_a = _calculate_chunk_offsets(a)
+    offsets_b = _calculate_chunk_offsets(b)
+    offsets = sorted(set(list(offsets_a) + list(offsets_b)))
+    in_a_offsets = _in_chunk_offsets(a, offsets)
+    in_b_offsets = _in_chunk_offsets(b, offsets)
+    return in_a_offsets, in_b_offsets
+
+
+def np_ufunc_op(op: Callable, a: Union[pa.ChunkedArray, pa.Array], b: Any):
```

Comment:
Are `ufunc`s always row-wise operations?

Reply:
Yes, ufuncs are by-default row-wise. They also need to support reductions but only via `ufunc.reduce` whereas `ufunc()` is always row-wise.